events/diskpersist.go

```diff
@@ -1,19 +1,447 @@
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/bluesky-social/indigo/api/atproto"
+	"github.com/bluesky-social/indigo/models"
+	"github.com/bluesky-social/indigo/util"
+	lru "github.com/hashicorp/golang-lru"
+	"gorm.io/gorm"
+)
+
 type DiskPersistence struct {
-	dir string
+	primaryDir    string
+	archiveDir    string
+	eventsPerFile int64
+
+	meta *gorm.DB
+
+	broadcast func(*XRPCStreamEvent)
+
+	logfi *os.File
+
+	curSeq int64
+	seqLk  sync.Mutex
+
+	uidCache *lru.ARCCache
+	didCache *lru.ARCCache
+
+	buffers *sync.Pool
+	scratch []byte
+}
+
+var _ (EventPersistence) = (*DiskPersistence)(nil)
+
+type DiskPersistOptions struct {
+	UIDCacheSize  int
+	DIDCacheSize  int
+	EventsPerFile int64
+}
+
+func DefaultDiskPersistOptions() *DiskPersistOptions {
+	return &DiskPersistOptions{
+		EventsPerFile: 10000,
+		UIDCacheSize:  100000,
+		DIDCacheSize:  100000,
+	}
+}
+
+func NewDiskPersistence(primaryDir, archiveDir string, db *gorm.DB, opts *DiskPersistOptions) (*DiskPersistence, error) {
+	if opts == nil {
+		opts = DefaultDiskPersistOptions()
+	}
+
+	uidCache, err := lru.NewARC(opts.UIDCacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create uid cache: %w", err)
+	}
+
+	didCache, err := lru.NewARC(opts.DIDCacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create did cache: %w", err)
+	}
+
+	db.AutoMigrate(&LogFileRef{})
+
+	bufpool := &sync.Pool{
+		New: func() any {
+			return new(bytes.Buffer)
+		},
+	}
+
+	dp := &DiskPersistence{
+		meta:          db,
+		primaryDir:    primaryDir,
+		archiveDir:    archiveDir,
+		buffers:       bufpool,
+		uidCache:      uidCache,
+		didCache:      didCache,
+		eventsPerFile: opts.EventsPerFile,
+		scratch:       make([]byte, headerSize),
+	}
+
+	if err := dp.resumeLog(); err != nil {
+		return nil, err
+	}
+
+	return dp, nil
+}
+
+type LogFileRef struct {
+	gorm.Model
+	Path     string
+	Archived bool
+	SeqStart int64
+}
+
+func (dp *DiskPersistence) resumeLog() error {
+	var lfr LogFileRef
+	if err := dp.meta.Order("seq_start desc").Limit(1).Find(&lfr).Error; err != nil {
+		return err
+	}
+
+	if lfr.ID == 0 {
+		// no files, start anew!
+		return dp.initLogFile()
+	}
+
+	fi, err := os.Open(filepath.Join(dp.primaryDir, lfr.Path))
+	if err != nil {
+		return err
+	}
+
+	seq, err := scanForLastSeq(fi, -1)
+	if err != nil {
+		return fmt.Errorf("failed to scan log file for last seqno: %w", err)
+	}
+
+	dp.curSeq = seq
+	dp.logfi = fi
+
+	return nil
+}
+
+func (dp *DiskPersistence) initLogFile() error {
+	if err := os.MkdirAll(dp.primaryDir, 0775); err != nil {
+		return err
+	}
+
+	p := filepath.Join(dp.primaryDir, "evts-0")
+	fi, err := os.Create(p)
+	if err != nil {
+		return err
+	}
+
+	if err := dp.meta.Create(&LogFileRef{
+		Path:     "evts-0",
+		SeqStart: 0,
+	}).Error; err != nil {
+		return err
+	}
+
+	dp.logfi = fi
+	dp.curSeq = 1
+	return nil
+}
+
+func (dp *DiskPersistence) swapLog(ctx context.Context) error {
```

Comment:
Should this rely on the called to lock the seq or should we lock the seq inside this function?

Reply:
If the seq needs to be locked by the caller, we should put a comment on this function to remind ourselves later